### PR TITLE
Improving discovery service

### PR DESCRIPTION
**Pull request by mazarmi**

### CODE REVIEW

**Comment by travisn** [comment]
we'll need to squash the commits before merging

File: README.md [hunk 1]
```diff
@@ -1,31 +1,51 @@
-# discovery.etcd.io
+# Discovery Service
+
+
+Comparison with *discovery.etcd.io* package:
+- backward compatible
+
+
+
+# Usage
```
Comment (travisn) [inline README.md]: Would be great to have this documentation filled in

File: README.md [hunk 2]
```diff
@@ -1,31 +1,51 @@
...
+This project extends (forks) the discovery.etcd.io's project.
 
-This code powers the public service at https://discovery.etcd.io. The API is
-documented in the [etcd clustering documentation](https://github.com/coreos/etcd/blob/master/Documentation/dev-internal/discovery_protocol.md#public-discovery-service).
 
 # Configuration
 
-The service has three configuration options, and can be configured with either
-runtime arguments or environment variables.
+The service has three configuration options, and can be configured with runtime arguments.
```
Comment (travisn) [inline README.md:16]: castled also supports env variables for all command line arguments. They follow the pattern of upper-case, adding a castle prefix, and converting dashes to underscores. For example, the parameter `--etcd-urls` would be the env var `CASTLE_ETCD_URLS`. See the method `setFlagsFromEnv` which would be simple to re-use in this repo if desired. But maybe we don't care about env vars or even docs if we aren't exposing the discovery service to the open-source world.

File: README.md [hunk 3]
```diff
@@ -1,31 +1,51 @@
...
+* `--etcd-urls`: comma separated list of backend etcd cluster. Default value is set to `"http://127.0.0.10:2379,http://127.0.0.11:2379,http://127.0.0.12:2379"`
+* `--discovery-addr`: address:port of the discovery service.
+* `--dns-addr`: address:port of the discovery service.
 
 ## Docker Container
 
 You may run the service in a docker container:
 
 ```
-docker pull quay.io/coreos/discovery.etcd.io
```
Comment (travisn) [inline README.md:20]: updated docker container commands?

File: build [hunk 4]
```diff
@@ -1,3 +1,3 @@
 #!/bin/sh -e
 
-go build -o bin/discovery github.com/coreos/discovery.etcd.io
+go build -o bin/discovery github.com/quantum/discoveryservice/cmd
```
Comment (travisn) [inline build]: i wonder if `discovery` is a better name than the longer `discoveryservice` for this repo

File: .gitignore [hunk 5]
```diff
@@ -1,3 +1,34 @@
...
+
+debug
+bin/
+tools/
+vendor/
+
+.vscode/
+.glide/
+gin-bin
+cmd/discovery
```
Comment (travisn) [inline .gitignore:34]: when you squash please make sure the discovery binary is not found in the git history

File: cmd/discovery.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	etcdURLs      string //comma separated list of etcd urls
+	discoveryAddr string
+	dnsAddr       string
+)
+
+// RootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "discovery",
+	Short: "discovery tool to bootstrap etcd nodes and help clients find them.",
+	Long:  `https://github.com/quantum/discovery`,
```
Comment (travisn) [inline cmd/discovery.go:34]: interesting, this is exactly the repository name suggested in another comment :)

File: cmd/discovery.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	Run: func(cmd *cobra.Command, args []string) {
+		etcdMachines := strings.Split(etcdURLs, ",")
+		handling.InitAndServe(discoveryAddr, etcdMachines)
+	},
+}
+
+func init() {
+	rootCmd.PersistentFlags().StringVar(&etcdURLs, "etcd-urls",
+		"http://127.0.0.10:2379,http://127.0.0.11:2379,http://127.0.0.12:2379",
```
Comment (travisn) [inline cmd/discovery.go]: why do we have defaults to 127.0.0.10, 11, and 12?

File: cmd/discovery.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+		"comma separated list of etcd listen URLs")
+	rootCmd.PersistentFlags().StringVar(&discoveryAddr, "discovery-addr", "http://127.0.0.1:8087",
+		"address:port of the discovery service")
+	rootCmd.PersistentFlags().StringVar(&dnsAddr, "dns-addr", "127.0.0.1:5300",
```
Comment (travisn) [inline cmd/discovery.go]: `discovery-addr` has an `http` prefix, but `dns-addr` does not. Should they be consistent?

File: handlers/home.go [hunk 9]
```diff
@@ -4,10 +4,10 @@ import (
 	"net/http"
 )
 
+// HomeHandler handles the home url of the discovery service
 func HomeHandler(w http.ResponseWriter, r *http.Request) {
-
 	http.Redirect(w, r,
-		"https://coreos.com/docs/cluster-management/setup/cluster-discovery/",
+		"https://github.com/quantum/discoveryservice",
```
Comment (travisn) [inline handlers/home.go]: another place potentially for `quantum/discovery`

File: handlers/member.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/gorilla/mux"
+	ctx "golang.org/x/net/context"
+)
+
+const (
+	// DefaultClientTimeout is the default timeout for etcd client
+	DefaultClientTimeout = 5 * time.Second
+)
+
+// MemberHandler can be used to query the latest etcd quorum members in a cleaner format compare to the standard etcd
```
Comment (travisn) [inline handlers/member.go]: can you describe why it is cleaner?

File: handlers/member.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+// format provided by GET "/{token:[a-f0-9]{32}}"
+func MemberHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	token := vars["token"]
+	log.Println("token: ", token)
+
+	cfg := client.Config{
+		Endpoints:               etcdMachines,
```
Comment (travisn) [inline handlers/member.go]: where is etcdMachines set? At that point should you just create and cache the etcdClient instead of re-creating a connection on every call?

File: handlers/member.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+		// However, etcd stores them as peer endpoints (which are used for communication between etcd peers).
+		// The only difference is in the port number. We don't rely on predefined ports, but only assume
+		// peer port number is always client-port+1
+		peerEndpoint := strings.Split(node.Value, "=")[1] //format: "member-name=scheme://peer-ip:peer-port"
+		u, err := url.Parse(peerEndpoint)
+		if err != nil {
+			return nil, err
+		}
+		peerSplits := strings.Split(u.Host, ":")
+		peerPort, err := strconv.Atoi(peerSplits[1])
```
Comment (travisn) [inline handlers/member.go:58]: the url type doesn't have a helper to extract the port?

File: handlers/new.go [hunk 13]
```diff
@@ -13,25 +14,22 @@ import (
...
 func init() {
-	newCounter = prometheus.NewCounterVec(
-		prometheus.CounterOpts{
-			Name: "endpoint_new_requests_total",
-			Help: "How many /new requests processed, partitioned by status code and HTTP method.",
-		},
-		[]string{"code", "method"},
-	)
-	prometheus.MustRegister(newCounter)
+	CurrentLeader.Set("127.0.0.10:2379")
```
Comment (travisn) [inline handlers/new.go]: hard-coded ip address

File: handlers/new.go [hunk 14]
```diff
@@ -44,17 +42,21 @@ func generateCluster() string {
...
 		Transport: client.DefaultTransport,
 		// set timeout per request to fail fast when the target endpoint is unavailable
 		HeaderTimeoutPerRequest: time.Second,
 	}
-
-	u, _ := url.Parse(etcdHost)
+	u, err := url.Parse(hostAddr)
+	if err != nil {
+		log.Println("u: ", u)
+		return
```
Comment (travisn) [inline handlers/new.go]: you don't need to return the error and fail the setup?

File: handlers/renew.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+// RenewHandler handles all PUT and GET requests related to the formation of cluster.
+// TODO: make it transactional by using etcd v3 apis
+func RenewHandler(w http.ResponseWriter, r *http.Request) {
+	// Receive list of current cluster members
+	var renewMsg RenewMsg
+	err := json.NewDecoder(r.Body).Decode(&renewMsg)
+	if err != nil {
+		fmt.Println("error: ", err)
```
Comment (travisn) [inline handlers/renew.go]: need to return an http error? looks like it will continue in an error state

File: handlers/renew.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+	kAPI := client.NewKeysAPI(c)
+
+	resp, err := kAPI.Get(ctx.Background(), clusterKey, &client.GetOptions{Recursive: true})
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		return
+	}
+	log.Println("resp: ", resp)
+	for _, node := range resp.Node.Nodes {
+		log.Println("previous node: ", node.Key)
```
Comment (travisn) [inline handlers/renew.go]: is this debug logging or is this necessary?

File: handlers/renew.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	currentMembers := renewMsg.Members
+	for _, member := range currentMembers {
+		log.Println("member to be refreshed or added as a part of renew: ", member)
+		key := path.Join(clusterKey, member.MemberID)
+		value := member.Name + "=" + member.PeerAddr
+		t := time.Duration(renewMsg.TTL) * time.Second
+		log.Printf("refreshing- K: %v | V: %v | ttl: %v", key, value, t)
```
Comment (travisn) [inline handlers/renew.go]: debug logging? If you keep it please spell out the K and V for log readability

File: handlers/renewmsg.go [hunk 18]
```diff
@@ -0,0 +1,25 @@
+package handlers
+
+import "time"
+
+// RenewMsg is used to refresh the current view of the cluster in discovery service
+type RenewMsg struct {
+	// ClusterID is the cluster's unique identifier
+	ClusterID string `json:"cluster-id"`
```
Comment (travisn) [inline handlers/renewmsg.go]: use camel case to be consistent with the json casing used throughout castle. For example, it would translate to clusterId, requestTime, and leaderName.

File: handlers/renewmsg.go [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+	// RequestTime specifies the request time at the leader side. (for logging/debugging)
+	RequestTime time.Time `json:"reqtime"`
+	// LeaderName is the name of the current leader that sends this message (for logging/debugging)
+	LeaderName string `json:"leader-name"`
+}
+
+// Member represents the required data for a single member to update
+type Member struct {
+	// ClusterID is the cluster's unique identifier
+	MemberID string `json:"member-id"`
```
Comment (travisn) [inline handlers/renewmsg.go]: similar casing comment. You shouldn't need the member prefix either except for memberId.

File: handlers/token.go [hunk 20]
```diff
@@ -10,39 +10,34 @@ import (
...
 		u := url.URL{
 			Scheme:   "http",
-			Host:     currentLeader.String(),
+			Host:     CurrentLeader.String(),
 			Path:     path.Join("v2", "keys", "_etcd", "registry", r.URL.Path),
 			RawQuery: r.URL.RawQuery,
 		}
 
+		if r.Method == "PUT" && !strings.Contains(strings.ToLower(u.RawQuery), "ttl") {
+			u.RawQuery = u.RawQuery + "&ttl=" + strconv.Itoa(DefaultTTL)
```
Comment (travisn) [inline handlers/token.go]: why do we have to append the default ttl to the url?

File: http/http.go [hunk 21]
```diff
@@ -22,17 +24,26 @@ func Setup(etcdHost, discHost string) {
...
 	// Only allow exact tokens with GETs and PUTs
-	r.HandleFunc("/{token:[a-f0-9]{32}}", handlers.TokenHandler).
-		Methods("GET", "PUT")
-	r.HandleFunc("/{token:[a-f0-9]{32}}/", handlers.TokenHandler).
-		Methods("GET", "PUT")
-	r.HandleFunc("/{token:[a-f0-9]{32}}/{machine}", handlers.TokenHandler).
-		Methods("GET", "PUT", "DELETE")
-	r.HandleFunc("/{token:[a-f0-9]{32}}/_config/size", handlers.TokenHandler).
-		Methods("GET")
+	r.HandleFunc("/{token:[a-f0-9]{32}}", handlers.TokenHandler).Methods("GET", "PUT")
```
Comment (travisn) [inline http/http.go]: why change the whitespace? it shows an unnecessary diff so it's better to follow the existing pattern.

File: glide.lock [hunk 22]
```diff
@@ -1,127 +1,85 @@
...
-  version: 20500e2abd0d1f4564a499e83d11d6c73cd58c27
-  subpackages:
-  - mem
-  - sftp
-- name: github.com/spf13/cast
-  version: e31f36ffc91a2ba9ddb72a4b6a607ff9b3d3cb63
-- name: github.com/spf13/jwalterweatherman
-  version: 33c24e77fb80341fe7130ee7c594256ff08ccc46
+  version: 454a56f35412459b5e684fd5ec0f9211b94f002a
+- name: github.com/skynetservices/skydns
```
Comment (travisn) [inline glide.lock]: you intended to break out all the dns changes to another PR, right?

File: handlers/member.go [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+	c, err := client.New(cfg)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+	}
+
+	key := path.Join("_etcd", "registry", token)
+	kAPI := client.NewKeysAPI(c)
+
+	resp, err := kAPI.Get(ctx.Background(), key, &client.GetOptions{Recursive: true, Quorum: true})
```
Comment (travisn) [inline handlers/member.go]: Ok, if the discovery service relies on the etcd store, how do we keep it stable if a member goes down? we don't have a castle cluster in global that can automate the failover for us.

**Review by travisn — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (mazarmi) [reply README.md]: I improved the documentation

File: README.md (hunk 2 above)
Comment (mazarmi) [reply README.md:16]: done

File: README.md (hunk 3 above)
Comment (mazarmi) [reply README.md:20]: haven't tested, but updated. I will test it as a part of global/aws task.

File: cmd/discovery.go (hunk 6 above)
Comment (mazarmi) [reply cmd/discovery.go:34]: Repo renamed to discovery

File: handlers/member.go (hunk 10 above)
Comment (mazarmi) [reply handlers/member.go]: done

File: handlers/member.go (hunk 11 above)
Comment (mazarmi) [reply handlers/member.go]: Now we create an etcdClient and pass it to the handlers package.

File: handlers/member.go (hunk 23 above)
Comment (mazarmi) [reply handlers/member.go]: This is a separate etcd backend. This etcd doesn't need to be dynamic. Even `discovery.etcd.io` has been working on a single static instance of backend etcd server for a few years! It's simply good enough for discovery use case.

File: handlers/member.go (hunk 12 above)
Comment (mazarmi) [reply handlers/member.go:58]: no

File: handlers/token.go (hunk 20 above)
Comment (mazarmi) [reply handlers/token.go]: I'll discuss it offline.

File: http/http.go (hunk 21 above)
Comment (mazarmi) [reply http/http.go]: It was more readable, reverted.

File: README.md [hunk 24]
```diff
@@ -1,43 +1,34 @@
...
-- `GET /cluster-id/`: returns a list of all members of the cluster with given cluster-id
-- `POST /renew`:
-
-# Discussion
-
-- This implementation provides security by obscurity.
-
-# References
-This project extends (forks) the discovery.etcd.io's project.
+- adding `members/CLUSTER_TOKEN` which returns a comma-separated list of *latest* members of the cluster.
```
Comment (travisn) [inline README.md]: i think the API format you had originally was more understandable to state the route first, then a description of it after

File: README.md [hunk 25]
```diff
@@ -1,43 +1,34 @@
...
 
 
 # Configuration
 
 The service has three configuration options, and can be configured with runtime arguments.
 
-* `--etcd-urls`: comma separated list of backend etcd cluster. Default value is set to `"http://127.0.0.10:2379,http://127.0.0.11:2379,http://127.0.0.12:2379"`
-* `--discovery-addr`: address:port of the discovery service.
-* `--dns-addr`: address:port of the discovery service.
+* `--etcd-urls` / `DISC_ETCD_URLS`: comma separated list of backend etcd cluster members. Default value is set to `"http://127.0.0.1:2379"`
```
Comment (travisn) [inline README.md:18]: Are these the same options as upstream, or did we add them?

File: cmd/discovery.go [hunk 26]
```diff
@@ -35,12 +51,28 @@ func init() {
...
 		"address:port of the discovery service")
+
+	// load the environment variables
+	setFlagsFromEnv(rootCmd.Flags())
 	rootCmd.MarkFlagRequired("etcd-urls")
 }
 
+func setFlagsFromEnv(flags *pflag.FlagSet) error {
+	flags.VisitAll(func(f *pflag.Flag) {
+		envVar := "DISC_" + strings.Replace(strings.ToUpper(f.Name), "-", "_", -1)
```
Comment (travisn) [inline cmd/discovery.go]: i think you should spell out the prefix: DISCOVERY

File: cmd/discovery.go [hunk 27]
```diff
@@ -22,7 +26,19 @@ var rootCmd = &cobra.Command{
 	Long:  `https://github.com/quantum/discovery`,
 	Run: func(cmd *cobra.Command, args []string) {
 		etcdMachines := strings.Split(etcdURLs, ",")
-		err := handling.InitAndServe(discoveryAddr, etcdMachines)
+		cfg := client.Config{
```
Comment (travisn) [inline cmd/discovery.go]: suggestion to split out this inline method to keep the cobra command looking simple

File: handlers/new.go [hunk 28]
```diff
@@ -24,14 +23,19 @@ const DefaultLeaderHost = "127.0.0.1:2379"
 var (
 	cfg           *client.Config
 	discHost      string
-	CurrentLeader lockstring.LockString
+	currentLeader lockstring.LockString
 	etcdMachines  []string
-	baseURI       = flag.String("host", "https://discovery.castle.io",
-		"base location for computed token URI")
+	baseURI       = flag.String("host", "https://discovery.castle.io", "base location for computed token URI")
+	etcdClient    client.KeysAPI
```
Comment (travisn) [inline handlers/new.go]: instead of a package var, what if you follow the pattern of the Handler type in castled/pkg/api/handlers.go?

**Review by travisn — APPROVED** [review]
